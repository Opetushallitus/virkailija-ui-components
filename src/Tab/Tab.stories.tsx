import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean } from '@storybook/addon-knobs';
import Tab from './index';

const stories = storiesOf('Components/Tab', module);

stories.add('Default', () => {
  const active = boolean('Active', false) as boolean;
  const disabled = boolean('Disabled', false) as boolean;

  return (
    <Tab active={active} disabled={disabled}>
      Tab
    </Tab>
  );
});
