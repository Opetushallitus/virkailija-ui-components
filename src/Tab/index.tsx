import * as React from 'react';
import styled, { css } from 'styled-components';

import { disabledStyle } from '../system';

const TabBase = styled.button.attrs({ role: 'tab' })<{
  active: boolean;
  disabled: boolean;
}>`
  outline: none;
  background-color: transparent;
  border: 0px none;
  cursor: pointer;
  transition: color 0.25s, border-color 0.25s;
  display: inline-flex;
  border-bottom: 3px solid transparent;
  padding: 6px 20px;

  ${({ theme }) => theme.typography.body};
  font-weight: ${({ theme }) => theme.fontWeights.bold};

  &:hover,
  &active {
    color: ${({ theme }) => theme.colors.primary.main};
  }

  &:focus {
    border-color: ${({ theme }) => theme.colors.primary.focusOutline};
  }

  ${disabledStyle}

  ${({ active }) =>
    active &&
    css`
      color: ${({ theme }) => theme.colors.primary.main};

      &,
      &:focus {
        border-color: ${({ theme }) => theme.colors.primary.main};
      }
    `};
`;

type TabBaseProps = {
  active?: boolean;
  value?: string;
  disabled?: boolean;
};

export type TabProps = Omit<
  React.ComponentProps<typeof TabBase>,
  keyof TabBaseProps
> &
  TabBaseProps;

const Tab = React.forwardRef<HTMLButtonElement, TabProps>(
  ({ active = false, disabled = false, ...props }, ref) => {
    return <TabBase active={active} disabled={disabled} ref={ref} {...props} />;
  },
);

export default Tab;
