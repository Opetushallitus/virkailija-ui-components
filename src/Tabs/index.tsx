import * as React from 'react';
import styled from 'styled-components';

import isFunction from '../utils/isFunction';

type TabsAlignTabs = 'start' | 'center' | 'end';

const mapAlignToJustifyContent = {
  start: 'flex-start',
  center: 'center',
  end: 'flex-end',
} as const;

const Wrapper = styled.div.attrs({ role: 'tablist' })<{
  alignTabs: TabsAlignTabs;
  fullWidth: boolean;
}>`
  display: ${({ fullWidth }) => (fullWidth ? 'flex' : 'inline-flex')};
  justify-content: ${({ alignTabs }) => mapAlignToJustifyContent[alignTabs]};
`;

export type TabsBaseProps = {
  value?: string;
  onChange?: (value: string) => void;
  fullWidth?: boolean;
  alignTabs?: TabsAlignTabs;
  children: React.ReactElement<{
    value?: string;
    active?: boolean;
    onClick?: (args: any) => void;
  }>[];
};

export type TabsProps = Omit<
  React.ComponentProps<typeof Wrapper>,
  keyof TabsBaseProps
> &
  TabsBaseProps;

const Tabs = ({
  value,
  onChange = () => {},
  alignTabs = 'start',
  fullWidth = true,
  ...props
}: TabsProps) => {
  const childrenCount = React.Children.count(props.children);

  const children = React.Children.map(props.children, (child, index) => {
    const active = value !== undefined && child.props.value === value;
    const isLast = index === childrenCount - 1;

    return React.cloneElement(child, {
      active,
      onClick: (args: any) => {
        isFunction(child.props.onClick) && child.props.onClick(args);
        child.props.value && onChange(child.props.value);
      },
    });
  });

  return (
    <Wrapper alignTabs={alignTabs} fullWidth={fullWidth} {...props}>
      {children}
    </Wrapper>
  );
};

export default Tabs;
