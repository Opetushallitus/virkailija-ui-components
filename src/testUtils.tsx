import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { createTheme } from './createTheme';

const theme = createTheme();

const Providers = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const dummyDiv = document.createElement('div');
dummyDiv.setAttribute('data-nothing', 'true');

// Return rootElement from render() for easier snapshot testing
const customRender = (ui: React.ReactElement, options: any = {}) => {
  const result = render(ui, { wrapper: Providers, ...options });
  return {
    ...result,
    rootElement: result.container.firstChild || dummyDiv,
  };
};

export * from '@testing-library/react';

export { customRender as render };
