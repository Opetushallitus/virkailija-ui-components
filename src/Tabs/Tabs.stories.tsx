import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, select } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

import Tabs from './index';
import Tab from '../Tab';

const change = action('change') as any;

const stories = storiesOf('Components/Tabs', module);

stories.add('Default', () => {
  const fullWidth = boolean('Full width', true) as boolean;

  const alignTabs = select(
    'Align tabs',
    {
      Start: 'start',
      Center: 'center',
      End: 'end',
    },
    'start',
  ) as any;

  return (
    <Tabs
      value="2"
      onChange={change}
      fullWidth={fullWidth}
      alignTabs={alignTabs}
    >
      <Tab value="1">Tab 1</Tab>
      <Tab value="2">Tab 2</Tab>
      <Tab value="3">Tab 3</Tab>
    </Tabs>
  );
});
