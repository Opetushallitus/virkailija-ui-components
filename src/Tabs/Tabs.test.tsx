import React from 'react';

import Tabs from './index';
import Tab from '../Tab';
import { render } from '../testUtils';

test('Renders correctly with 3 tabs', () => {
  const { rootElement } = render(
    <Tabs value="b">
      <Tab value="a">Tab A</Tab>
      <Tab value="b">Tab B</Tab>
      <Tab value="c">Tab C</Tab>
    </Tabs>,
  );
  expect(rootElement).toMatchSnapshot();
});
