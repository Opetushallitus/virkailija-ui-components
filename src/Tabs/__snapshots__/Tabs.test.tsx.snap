// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly with 3 tabs 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c1 {
  outline: none;
  background-color: transparent;
  border: 0px none;
  cursor: pointer;
  -webkit-transition: color 0.25s,border-color 0.25s;
  transition: color 0.25s,border-color 0.25s;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  border-bottom: 3px solid transparent;
  padding: 6px 20px;
  line-height: 1.5;
  font-family: 'Roboto',sans-serif;
  font-size: 1rem;
  color: #666;
  font-weight: 500;
}

.c1:hover,
.c1active {
  color: #0a789c;
}

.c1:focus {
  border-color: rgba(10,120,156,0.2);
}

.c2 {
  outline: none;
  background-color: transparent;
  border: 0px none;
  cursor: pointer;
  -webkit-transition: color 0.25s,border-color 0.25s;
  transition: color 0.25s,border-color 0.25s;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  border-bottom: 3px solid transparent;
  padding: 6px 20px;
  line-height: 1.5;
  font-family: 'Roboto',sans-serif;
  font-size: 1rem;
  color: #666;
  font-weight: 500;
  color: #0a789c;
}

.c2:hover,
.c2active {
  color: #0a789c;
}

.c2:focus {
  border-color: rgba(10,120,156,0.2);
}

.c2,
.c2:focus {
  border-color: #0a789c;
}

<div
  class="c0"
  role="tablist"
>
  <button
    class="c1"
    role="tab"
    value="a"
  >
    Tab A
  </button>
  <button
    class="c2"
    role="tab"
    value="b"
  >
    Tab B
  </button>
  <button
    class="c1"
    role="tab"
    value="c"
  >
    Tab C
  </button>
</div>
`;
